skimage/color/_colorconv.pyx

```diff
@@ -0,0 +1,83 @@
+#cython: boundscheck=False
+#cython: wraparound=False
+#cython: cdivision=True
+
+import cython
+import numpy as np
+from libc.math cimport floor
+
+from .._shared.fused_numerics cimport np_floats
+
+
+def rgb2hsv_inner(np_floats[:, ::] rgb):
+    cdef:
+        np_floats[:, ::] hsv = np.empty_like(rgb)
+        np_floats delta, minv, maxv, tmp
+        Py_ssize_t i, n, ch
+
+    n = rgb.shape[0]
+    with nogil:
+        for i in range(n):
+            minv = maxv = rgb[i, 0]
+            for ch in range(1, 3):
+                tmp = rgb[i, ch]
+                if tmp > maxv:
+                    maxv = tmp
+                elif tmp < minv:
+                    minv = tmp
+            delta = maxv - minv
+            if delta == 0.0:
+                hsv[i, :2] = 0.0
+            else:
+                hsv[i, 1] = delta / maxv
+                if rgb[i, 0] == maxv:
+                    hsv[i, 0] = (rgb[i, 1] - rgb[i, 2]) / delta
+                elif rgb[i, 1] == maxv:
+                    hsv[i, 0] = 2.0 + (rgb[i, 2] - rgb[i, 0]) / delta
+                elif rgb[i, 2] == maxv:
+                    hsv[i, 0] = 4.0 + (rgb[i, 0] - rgb[i, 1]) / delta
+                hsv[i, 0] /= 6.0
+                hsv[i, 0] -= floor(<double>hsv[i, 0])
```

Comment:
I am not sure that this will work but what about defining a `_floor` function specific to `np_floats`?
```
if np_floats is cnp.float32_t:
    _floor = floorf
else:
    _floor = floor
```
it may avoid this cast to double precision when the input is single precision...

Reply:
I think we would need to a C99 flag (or `language='c++'`) to ensure `floorf` is available. I think that used to be a problem for older MSVC versions, but is probably fine now. 

Cython 0.29.x doesn't define floorf in it's `libc.math`, but the current 3.0 alpha does. That doesn't really matter, though, since we can always just add it ourselves with: 
```Cython
cdef extern from "<math.h>" nogil:
    float floorf(float)
```